src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerArgumentBinder.cs

```diff
@@ -163,10 +163,11 @@ public class ControllerArgumentBinder : IControllerArgumentBinder
                 parameter.BindingInfo,
                 parameter.Name);
 
-            if (parameter.BindingInfo?.BinderModelName != null)
+            var parameterModelName = parameter.BindingInfo?.BinderModelName ?? metadata.BinderModelName;
```

Comment:
> The `TryUpdateModel` code in `ModelBindingHelper` applies the same logic and will need to be updated

Turns out no change is needed in `ModelBindingHelper` because it doesn't really have special cases for top-level expressions. The only use of `MemberInfo` or subclasses there is to determine C# property names.